### PR TITLE
am2r: init at 29

### PR DESCRIPTION
**Pull request by IvarWithoutBones**
###### Description of changes
This adds AM2R, a fan-made remake of Metroid II: Return of Samus in the style of Metroid: Zero Mission. I've also added a multiplayer mod and a server for it. See https://github.com/AM2R-Community-Developers/AM2R-Autopatcher-Linux for the utility patching it with Linux support, and https://github.com/AM2R-Community-Developers/AM2R-Community-Updates for the community updates.

A multiplayer mod and a server for it are added as well. Tested it locally and everything seems to work flawlessly :)

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tpwrules** [comment]
It looks like these should have a `sourceProvenance` attribute set appropriately as there's lots of binary code.

**Comment by Miepee** [comment]
@SuperSamus coming from your original PR, a few notes:
- If you (or anyone else) still want to package the AM2RLauncher for nix, feel free to do so, as I've added the option for it to not create AppImages via a compile flag.
- The resuling binaries not running may have to do something with the fact that YoYoGames wrote them a little weird (and the different Multitroid version being compiled differently). They require some environment variables for Mesa in order to run correctly (something that this PR has done)
- If you want to package software for something, it can't hurt to give a small heads up to to project maintainers. We might've been help and prevented you loosing the motivation for it.

---
Also, as a more general question, would it make sense to publish both the AM2RLauncher, and the main game as one package, seperately? I feel like it's kinda duplicated effort and can get messy (especially if multiple mods get packaged in there as well), but can also see why someone would not want to download the .NET runtime just to play a small game.
Am also not a nix user so no idea what the community prefers.

**Comment by SuperSamus** [comment]
@Miepee I did actually update the branch to fix the Mesa problem, but PRs don't update any more once closed, meaning that the correct link to look is [this](https://github.com/SuperSamus/nixpkgs/tree/am2r).

About the AM2R launcher. The problem isn't the AppImage per se, it's the fact that NixOS doesn't respect FHS, meaning that dependencies need to be either (simplified):
- Linked during the build. Nix does it by itself, but AM2RLauncher doesn't ([yet](https://github.com/AM2R-Community-Developers/AM2RLauncher/blob/main/AM2RLauncher/AM2RLauncherLib/Profile.cs#L483)).
- Put in a sandbox, in order to create a FHS environment for them.

Meaning that in order to package AM2RLauncher what needs to be done is:
- Create a FHS environment for the old 32 bit libraries, some of them I think aren't even packaged here anymore (I don't think it's possible to include libraries that aren't packaged). (Not necessary if AM2RLauncher starts using `patchelf`)
- Patch the launcher to launch the AppImage with `appimage-run` (which isn't the lightest dependency). Also, I never tested whether AM2R's AppImage works with `appimage-run`.
- If one of them is impossible, patch AM2R launcher to disable the option to (not) use an AppImage (to force using the one that works).

In short... I don't think it's worth it.

Anyway, mods shouldn't need any intervention (aside from cases like Multitroid which require an extra application), and they don't even need to be included on nixpkgs: a user could simply do something like:

```nix
environment.systemPackages = [
    (am2r.override { customClient = /path/to/mod };)  # You can also use builtins.fetchurl and stuff
]
```

(In my version, I also made the executable name dependent on the package name, meaning that it could be overwritten in order to allow for multiple mods to be installed. This PR doesn't. (It's one of the things I meant with "maybe you could take something from it".))

**Comment by IvarWithoutBones** [comment]
Apologies for the late reply! I currently don't have the time to work on this, so absolutely feel free to open your own PR in favour of this one. It'll most likely get merged quicker as I imagine i wont be able to address the feedback any time soon.

One bit of feedback I have is that it would be preferable to use the official tooling rather than doing all the logic manually inside of the derivation. The author of the official script is interested in this as well, so potential flaws with it would be best fixed upstream IMO.
